### PR TITLE
git integration

### LINKED ISSUE #43: Use git to crawl for dependencies
Instead of reading and downloading data from GitHub and its API, use `git clone` which allows the utilization of any Git repository. 

This will:
- clone the repositories (only lockfiles with `sparse-checkout`)
- read from the lock files
- ~~remove the asynchronous things from the lock file parser due to weird behavior~~ (done in #73)

### PR DESCRIPTION
**Pull request by c0rydoras**
Resolves #43

Possible improvements:
- parser
  - use `poetry` directly to parse and interpret lockfiles instead of `tomllib`
  - use [pyarn](https://pypi.org/project/pyarn/) for yarn.lock instead of regex as its tested
  - add `yarn@^3.0.0` parser, this should be easy as modern yarn.lock files are yaml 
  - add support for `yarn.lock` past yarn classic. this does not require an extra library as its yaml
  - use `dependencies` and `devDependencies` keys for `pnpm-lock.yaml` instead of packages and regex

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -26,7 +26,7 @@ api-start: ## Start the API
 
 .PHONY: api-test
 api-test: ## Test the API
-	@docker compose run --rm api pytest --no-cov-on-fail --cov -vvv -s
+	@docker compose run --rm api pytest --no-cov-on-fail --cov -vvv --create-db -s
```
Comment (open-dynaMIX) [inline Makefile:29]: Did you mean to write `--reuse-db`?

**Review by open-dynaMIX — CHANGES_REQUESTED** [review]
Great work :rocket:

File: api/outdated/models.py [hunk 2]
```diff
@@ -57,11 +59,25 @@ def pre_save(self, model_instance, add):
...
+    """Validate the existance of a remote git repository."""
+    result = run(
+        ["/usr/bin/git", "ls-remote", "https://" + value],
+        capture_output=True,
+        check=False,
+        shell=False,
+    )
+    if result.returncode == 0:
+        return
+    raise ValidationError("Repository does not exist.", params={"value": value})
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/models.py]: fixed here: https://github.com/adfinis/Outdated/pull/141/commits/8178c3540c86fed0047f13d02c86815a1fbc6da4

File: api/outdated/tests/test_repo_field.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+from django.core.exceptions import ValidationError
+
+from outdated.models import validate_repo_exists
+
+
+@pytest.mark.parametrize(
+    "repo,exists",
+    [("github.com/adfinis/outdated", True), ("adfinis.com/adfinis/outdated", False)],
+)
+def test_repository_exists_validator(repo: str, exists: bool) -> None:
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/tests/test_repo_field.py]: i now use a local git repository with `file://`, changed here: https://github.com/adfinis/Outdated/pull/141/commits/8178c3540c86fed0047f13d02c86815a1fbc6da4

File: Makefile (hunk 1 above)
Comment (c0rydoras) [reply Makefile]: `--reuse-db` is in the pyproject.toml file (pytest.addopts) i removed `--create-db` from the makefile here: https://github.com/adfinis/Outdated/pull/141/commits/512ac4725ab1cc163c1e5ac19418711bdd58ba29

File: api/outdated/conftest.py [hunk 4]
```diff
@@ -57,3 +67,30 @@ def client(db, settings, get_claims):
...
+            tracker_sync_mock = tracker_mock('sync')
+            Tracker(project).sync()
+            tracker_sync_mock.assert_called_once()
+        ```
+    """
+
+    def _tracker_mock(target: str) -> MagicMock:
+        return mocker.patch.object(Tracker, target)
+
+    return _tracker_mock
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply api/outdated/conftest.py:92]: Its not necessary, but I prefer it this way and as its less to type

File: api/outdated/models.py [hunk 5]
```diff
@@ -57,11 +60,29 @@ def pre_save(self, model_instance, add):
         return super().pre_save(model_instance, add)
 
 
+def validate_repo_exists(value: str) -> None:
+    """Validate the existance of a remote git repository."""
+    url = "https://" + value
+
+    if value.startswith("file://") and settings.ENV == "test":
+        url = value
```
Comment (open-dynaMIX) [inline api/outdated/models.py:68]: What's the reasoning behind not storing the scheme in the DB?

Comment (c0rydoras) [reply api/outdated/models.py:68]: Repos without ssl are out of scope, and `file://` will only ever be used in tests

See:
- https://github.com/adfinis/Outdated/issues/91 
- https://github.com/adfinis/Outdated/issues/93
- https://github.com/adfinis/Outdated/pull/92#issuecomment-1715624976

Comment (open-dynaMIX) [reply api/outdated/models.py:68]: I see. Fine by me :+1: Just as an input, if you plan to extend this in the future: If you store the scheme in a different model field, validation for uniqueness won't be more difficult than now.

**Review by open-dynaMIX — APPROVED** [review]

**Pull request merged**